timApp/auth/oauth2/routes.py

```diff
@@ -65,3 +86,127 @@ def get_user_profile() -> Response:
             "username": user.name,
         }
     )
+
+
+@oauth.get("bookmarked-courses")
+@require_oauth(Scope.user_courses.value)
+def get_courses() -> Response:
+    """
+    Gets all courses that the user has bookmarked
+    :return: JSON response with all courses TODO: Remove if not needed
+    """
+    user: User = current_token.user
+    courses = user.bookmarks.bookmark_data[2]["My courses"]
+
+    res = [
+        {"course_name": list(course.keys())[0], "path": list(course.values())[0]}
+        for course in courses
+    ]
+
+    return json_response(res)
+
+
+@oauth.get("ide-courses")
+@require_oauth(Scope.user_tasks.value)
+def get_user_ide_courses() -> Response:
+    """
+    Get all courses that the user has bookmarked and have ideCourse tah
+    :return: JSON response with all courses and their demo folders
+    """
+
+    user: User = current_token.user
+    return json_response(user_ide_courses(user=user))
+
+
+@oauth.get("demos-by-doc-id")
+@require_oauth(Scope.user_tasks.value)
+def get_ide_demos_by_doc_id() -> Response:
+    """
+    Get all demos by document id
+    :return: JSON response with the task
+    """
+
+    doc_id = request.json.get("doc_id")
```

Comment:
Meillä polkuja varten on "erikoinen" [TypedBlueprint](https://tim.jyu.fi/view/tim/TIMin-kehitys/Johdatus-TIMin-kehitykseen#flaskin-reitit), jonka avulla kaikki nuo request.json yms voi laittaa suoraan polkufunktion parametrikisi. Eli siis tuollainen on mahdollinen ja sallittu

```suggestion
@oauth.get("demos-by-doc-id")
@require_oauth(Scope.user_tasks.value)
def get_ide_demos_by_doc_id(doc_id: int) -> Response:
    """
    Get all demos by document id
    :return: JSON response with the task
    """
```

Tämän hyödyt:

- doc_id on nyt pakko olla int
- doc_id on nyt pakko olla määritelty (se voi olla vapaaehtoinen, jos antaa argumentille oletusarvon)

"Haitat":

- GET-poluissa argumentit otetaan URL-parametrista (eli siis `/oauth/demos-by-doc-id?doc_id=1`) eikä se vielä tue arvon nappaamista JSONista
- POST-polussa argumentit otetaan JSONista eikä se vielä tue arvon nappaamista URL-parametrista

Sanoisin, että kuitenkin tuo on varsin kätevä tapa määritellä polulle parametreja.